### PR TITLE
Added tests for editing spectrum title and notes

### PR DESCRIPTION
**Pull request by waridrox**
#699 referenced the missing routes for the edit and update spectrum. Once that is fixed, these tests will secure the functionality for editing and updating the spectrum title and notes.

Screenshot 2021-08-16 at 6 24 13 PM

* [ ] tests pass -- `rake test`
* [ ] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [ ] pull request are descriptively named
* [ ] if possible, multiple commits squashed if they're smaller changes
* [ ] reviewed/confirmed/tested by another contributor or maintainer

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/wiki/contributing-to-public-lab-software

Thanks!

### CODE REVIEW

**Comment by jywarren** [comment]
Hi @waridrox would you like me to rebase this? Thanks!

**Comment by waridrox** [comment]
Hmm...I think the functionality to edit and update the title / notes of the spectrum is still non-functional on `spectral-workbench.org`.

**Comment by jywarren** [comment]
Is it another routes issue? Let's try adding those routes explicitly. That's the more correct way in recent Rails.

**Comment by jywarren** [comment]
Hmm. Those routes should've been added here:

https://github.com/waridrox/spectral-workbench/commit/8ce0c273e6cccb910dbd634e63292a86db66a182#diff-959bc9abc46a55332bb64d5155a79323afa75a50ec1a2137ddd22d926f62c6c5R43-R48
Errors are:

```
Failure:
SpectrumsControllerTest#test_should_update_spectrum_notes [/home/runner/work/spectral-workbench/spectral-workbench/test/controllers/spectrums_controller_test.rb:264]:
Expected: "Neon spectrum info"
  Actual: "A test spectrum."

bin/rails test test/controllers/spectrums_controller_test.rb:258

........................F

Failure:
SpectrumsControllerTest#test_should_update_spectrum_title [/home/runner/work/spectral-workbench/spectral-workbench/test/controllers/spectrums_controller_test.rb:252]:
Expected: "Really cool spectrum"
  Actual: "One nice spectrum"

bin/rails test test/controllers/spectrums_controller_test.rb:246

....................
```

**Comment by jywarren** [comment]
That's great - confirmed in GitPod that saving doesn't work, but it's not routes. So these tests are critical and exposing an existing bug - perfect use case for tests!!

Looking into why now...

File: test/controllers/spectrums_controller_test.rb [hunk 1]
```diff
@@ -243,6 +243,29 @@ class SpectrumsControllerTest < ActionController::TestCase
...
+    # to check if the new spectrum title was updated
+    assert_equal "Really cool spectrum" , spectrums(:one).title
+    assert_not_nil :spectrum
+    assert_redirected_to spectrum_path(assigns(:spectrum))
+  end
+
+
+  test 'should update spectrum notes' do
+    session[:user_id] = User.first.id # log in
+    patch :update, params: { id: spectrums(:one).id, spectrum: { notes: "Neon spectrum info." } }
```
Comment (jywarren) [inline test/controllers/spectrums_controller_test.rb:260]: OK, log upon doing this action manually in gitpod:

```


Started PATCH "/spectrums/1" for 10.68.6.9 at 2021-09-05 17:49:50 +0000
Processing by SpectrumsController#update as HTML
  Parameters: {"utf8"=>"✓", "authenticity_token"=>"nQMzQNOU8GZimN+bCpA7BqdOqJe5yLFypr79cd9iUYtf7OpoJJO9bu3ncSU2MsT30XvCGjDpkNkJvlj4kkaooA==", "spectrum"=>{"title"=>"test 3", "notes"=>""}, "id"=>"1"}
  User Load (0.7ms)  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 1 LIMIT 1
  ↳ app/controllers/application_controller.rb:31
  CACHE User Load (0.0ms)  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 1 LIMIT 1  [["id", 1], ["LIMIT", 1]]
  ↳ app/controllers/application_controller.rb:87
  Spectrum Load (0.7ms)  SELECT  `spectrums`.* FROM `spectrums` WHERE `spectrums`.`id` = 1 LIMIT 1
  ↳ app/controllers/spectrums_controller.rb:349
  CACHE User Load (0.0ms)  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 1 LIMIT 1  [["id", 1], ["LIMIT", 1]]
  ↳ app/controllers/application_controller.rb:87
  CACHE User Load (0.0ms)  SELECT  `users`.* FROM `users` WHERE `users`.`id` = 1 LIMIT 1  [["id", 1], ["LIMIT", 1]]
  ↳ app/controllers/application_controller.rb:31
   (0.3ms)  BEGIN
  ↳ app/controllers/spectrums_controller.rb:358
  Tag Load (0.5ms)  SELECT `tags`.* FROM `tags` WHERE `tags`.`spectrum_id` = 1
  ↳ app/models/spectrum.rb:185
   (0.4ms)  COMMIT
  ↳ app/controllers/spectrums_controller.rb:358
Redirected to https://3000-silver-thrush-xcnyep74.ws-us16.gitpod.io/spectrums/1
Completed 302 Found in 18ms (ActiveRecord: 2.6ms)
```

File: test/controllers/spectrums_controller_test.rb [hunk 2]
```diff
@@ -243,6 +243,29 @@ class SpectrumsControllerTest < ActionController::TestCase
...
+
+    assert_response :redirect
+    # to check if the new spectrum notes was updated
+    assert_equal "Neon spectrum info", spectrums(:one).notes
```
Comment (jywarren) [inline test/controllers/spectrums_controller_test.rb:264]:
```suggestion
    assert_equal flash[:success], 'Spectrum was successfully updated.'
    assert_equal "Neon spectrum info", spectrums(:one).notes
```

File: test/controllers/spectrums_controller_test.rb [hunk 3]
```diff
@@ -243,6 +243,29 @@ class SpectrumsControllerTest < ActionController::TestCase
     assert_redirected_to spectrum_path(assigns(:spectrum))
   end
 
+  test 'should update spectrum title' do
+    session[:user_id] = User.first.id # log in
+    patch :update, params: { id: spectrums(:one).id, spectrum: { title: "Really cool spectrum" } }
+
+    assert_response :redirect
+    # to check if the new spectrum title was updated
+    assert_equal "Really cool spectrum" , spectrums(:one).title
```
Comment (jywarren) [inline test/controllers/spectrums_controller_test.rb:252]:
```suggestion
    assert_equal flash[:success], 'Spectrum was successfully updated.'
    assert_equal "Really cool spectrum" , spectrums(:one).title
```

File: test/controllers/spectrums_controller_test.rb [hunk 4]
```diff
@@ -243,6 +243,29 @@ class SpectrumsControllerTest < ActionController::TestCase
     assert_redirected_to spectrum_path(assigns(:spectrum))
   end
 
+  test 'should update spectrum title' do
+    session[:user_id] = User.first.id # log in
+    patch :update, params: { id: spectrums(:one).id, spectrum: { title: "Really cool spectrum" } }
```
Comment (jywarren) [inline test/controllers/spectrums_controller_test.rb:248]: aha -- ok, so this actually is being looked for as simply `title` not `spectrum: { title: "X" }`:

https://github.com/waridrox/spectral-workbench/blob/a105a5c1db55c37cbd8efa094217a48f56bb91a5/app/controllers/spectrums_controller.rb#L352-L358

However even from the user interface we're sending: 

`  Parameters: {"utf8"=>"✓", "authenticity_token"=>"nQMzQNOU8GZimN+bCpA7BqdOqJe5yLFypr79cd9iUYtf7OpoJJO9bu3ncSU2MsT30XvCGjDpkNkJvlj4kkaooA==", "spectrum"=>{"title"=>"test 3", "notes"=>""}, "id"=>"1"}`

Let's update the controller.

File: app/controllers/spectrums_controller.rb [hunk 5]
```diff
@@ -349,9 +349,9 @@ def update
     @spectrum = Spectrum.find(params[:id])
     require_ownership(@spectrum)
 
-    @spectrum.title = params[:title] unless params[:title].nil?
-    @spectrum.notes = params[:notes] unless params[:notes].nil?
-    @spectrum.data  = params[:data] unless params[:data].nil?
+    @spectrum.title = params[:spectrum][:title] unless params[:spectrum].nil? || params[:spectrum][:title].nil?
+    @spectrum.notes = params[:spectrum][:notes] unless params[:spectrum].nil? || params[:spectrum][:notes].nil?
+    @spectrum.data  = params[:spectrum][:data] unless  params[:spectrum].nil? || params[:data].nil?
```
Comment (jywarren) [inline app/controllers/spectrums_controller.rb:354]: Wasn't sure where we allow this, actually. not from the HTML UI form... but maybe via the API? Anyways i updated it to match.

Comment (jywarren) [reply app/controllers/spectrums_controller.rb:354]:
```
Failure:
SpectrumsControllerTest#test_should_update_spectrum_notes [/home/runner/work/spectral-workbench/spectral-workbench/test/controllers/spectrums_controller_test.rb:264]:
Expected: "Neon spectrum info"
  Actual: "A test spectrum."

bin/rails test test/controllers/spectrums_controller_test.rb:258

.........F

Failure:
SpectrumsControllerTest#test_should_update_spectrum_title [/home/runner/work/spectral-workbench/spectral-workbench/test/controllers/spectrums_controller_test.rb:252]:
Expected: "Really cool spectrum"
  Actual: "One nice spectrum"

bin/rails test test/controllers/spectrums_controller_test.rb:246
```

**Review by jywarren — COMMENTED** [review]
add flash assertions too

File: test/controllers/spectrums_controller_test.rb [hunk 6]
```diff
@@ -243,6 +243,31 @@ class SpectrumsControllerTest < ActionController::TestCase
...
+  end
+
+
+  test 'should update spectrum notes' do
+    session[:user_id] = User.first.id # log in
+    patch :update, params: { id: spectrums(:one).id, spectrum: { notes: "Neon spectrum info." } }
+
+    assert_response :redirect
+    # to check if the new spectrum notes was updated
+    assert_equal flash[:success], 'Spectrum was successfully updated.'
```
Comment (jywarren) [inline test/controllers/spectrums_controller_test.rb]: Flash assertions not working!

```
-"Spectrum was successfully updated."
+nil
```

File: test/controllers/spectrums_controller_test.rb [hunk 7]
```diff
@@ -243,6 +243,31 @@ class SpectrumsControllerTest < ActionController::TestCase
     assert_redirected_to spectrum_path(assigns(:spectrum))
   end
 
+  test 'should update spectrum title' do
+    session[:user_id] = User.first.id # log in
+    patch :update, params: { id: spectrums(:one).id, spectrum: { title: "Really cool spectrum" } }
+
+    assert_response :redirect
+    # to check if the new spectrum title was updated
+    assert_equal 'Spectrum was successfully updated.', flash[:success]
```
Comment (jywarren) [inline test/controllers/spectrums_controller_test.rb:252]: Flash assertions not working!

```
-"Spectrum was successfully updated."
+nil
```

**Comment by jywarren** [comment]
Fingers 🤞

**Comment by jywarren** [comment]
hooray!!!

**Pull request merged**